Delineate Anything Inference

isaaccorley opened the pull request:
This PR adds the ability to perform downloading of a single image and performing inference using the Delineate Anything models.

The new cli calls are `ftw inference run-instance-segmentation` which performs inference on an already downloaded image and `ftw inference ftw-inference-instance-segmentation-all` which combines

Note that this model is instance segmentation so there is no intermediate mask tif created, the results are converted directly to polygons in fiboa parquet format.

Some examples of the usage:

```shell
ftw inference ftw-inference-instance-segmentation-all S2B_MSIL2A_20210617T100559_R022_T33UUP_20210624T063729 --bbox 13.0,48.0,13.3,48.3  --model DelineateAnything-S --gpu 0 --out_dir outputs
```

```shell
ftw inference run-instance-segmentation inference_imagery/austria_example.tif --model DelineateAnything-S --gpu 0 --out tmp.parquet --patch_size 512 --image_size 1024 --conf_threshold=0.1 --iou_threshold=0.6 --min_size 100 --simplify 15
```

Results are decent, the post-processing and patch size parameters could probably use some tuning though 
<img width="932" height="561" alt="image" src="https://github.com/user-attachments/assets/a28163fc-57e5-413b-a784-b3e39d038829" />

m-mohr commented on ftw_tools/cli.py:
```diff
@@ -502,6 +511,13 @@ def inference_download(win_a, win_b, out, overwrite, bbox, stac_host):
 )
+@click.option(
+    "--num_workers",
+    type=int,
```
```suggestion
    type=click.IntRange(min=1),
```

m-mohr commented on ftw_tools/cli.py:
```diff
@@ -536,12 +553,351 @@ def inference_run(
+    "--model",
+    "-m",
+    type=click.Choice(["DelineateAnything", "DelineateAnything-S"]),
+    required=True,
```
Do we want to set a default for the lazy people? :-)

m-mohr commented on ftw_tools/cli.py:
```diff
@@ -536,12 +553,351 @@ def inference_run(
+@click.option(
+    "--out",
+    "-o",
+    type=str,
```
```suggestion
    type=click.Path(exists=False),
```

m-mohr commented on ftw_tools/cli.py:
```diff
@@ -536,12 +553,351 @@ def inference_run(
+    show_default=True,
+    help="Image size to use for inference.",
+)
+@click.option(
+    "--patch_size",
+    type=int,
+    default=None,
+    help="Size of patch to use for inference. Defaults to 1024 unless the image is < 1024x1024px.",
+)
```
```suggestion
@click.option(
    "--patch_size",
    "-ps",
    type=click.IntRange(min=128),
    default=None,
    help="Size of patch to use for inference. Defaults to 1024 unless the image is < 1024x1024px and a smaller value otherwise.",
)
```

Maybe check whether that's actually what you are implementing for <1024x1024...

m-mohr commented on ftw_tools/cli.py:
```diff
@@ -536,12 +553,351 @@ def inference_run(
+)
+@click.option(
+    "--num_workers",
+    type=int,
+    default=2,
```
```suggestion
    type=click.IntRange(min=1),
    default=4,
```

m-mohr commented on ftw_tools/cli.py:
```diff
@@ -536,12 +553,351 @@ def inference_run(
+)
+@click.option(
+    "--max_detections",
+    type=int,
```
```suggestion
    type=click.IntRange(min=1),
```

m-mohr commented on ftw_tools/cli.py:
```diff
@@ -536,12 +553,351 @@ def inference_run(
+    show_default=True,
+    help="Maximum number of detections to keep.",
+)
+@click.option(
+    "--iou_threshold",
+    type=float,
+    default=0.6,
+    show_default=True,
+    help="IoU threshold for matching detections to ground truths.",
+)
```
```suggestion
@click.option(
    "--iou_threshold",
    "-iou",
    type=click.FloatRange(min=0.0, max=1.0),
    default=0.5,
    show_default=True,
    help="IoU threshold for matching predictions to ground truths",
)
```

m-mohr commented on ftw_tools/cli.py:
```diff
@@ -536,12 +553,351 @@ def inference_run(
+)
+@click.option(
+    "--conf_threshold",
+    type=float,
```
```suggestion
    "-ct",
    type=click.FloatRange(min=0.0, max=1.0),
```

m-mohr commented on ftw_tools/cli.py:
```diff
@@ -536,12 +553,351 @@ def inference_run(
+)
+@click.option(
+    "--padding",
+    type=int,
```
```suggestion
    "-p",
    type=click.IntRange(min=0),
```

m-mohr commented on ftw_tools/cli.py:
```diff
@@ -536,12 +553,351 @@ def inference_run(
+    default=None,
+    help="Pixels to discard from each side of the patch.",
+)
+@click.option(
+    "--overwrite", "-f", is_flag=True, help="Overwrite outputs if they exist."
+)
```
```suggestion

@click.option(
    "--overwrite",
    "-f",
    is_flag=True,
    default=False,
    show_default=True,
    help="Overwrites the outputs if they exist",
)
```

m-mohr commented on ftw_tools/cli.py:
```diff
@@ -536,12 +553,351 @@ def inference_run(
+    show_default=True,
+    help="Simplification factor to use when polygonizing in the unit of the CRS, e.g. meters for Sentinel-2 imagery in UTM. Set to 0 to disable simplification.",
+)
+@click.option(
+    "--min_size",
+    type=float,
+    default=500,
+    show_default=True,
+    help="Minimum area size in square meters to include in the output. Set to 0 to disable.",
+)
```
```suggestion
@click.option(
    "--min_size",
    "-sn",
    type=click.FloatRange(min=0.0),
    default=500,
    show_default=True,
    help="Minimum area size in square meters to include in the output. Set to 0 to disable.",
)
```

m-mohr commented on ftw_tools/cli.py:
```diff
@@ -536,12 +553,351 @@ def inference_run(
+    show_default=True,
+    help="Minimum area size in square meters to include in the output. Set to 0 to disable.",
+)
+@click.option(
+    "--max_size",
+    type=float,
+    default=None,
+    show_default=True,
+    help="Maximum area size in square meters to include in the output. Disabled by default.",
+)
```
```suggestion
@click.option(
    "--max_size",
    "-sx",
    type=click.FloatRange(min=0.0),
    default=None,
    show_default=True,
    help="Maximum area size in square meters to include in the output. Disabled by default.",

```

m-mohr commented on ftw_tools/cli.py:
```diff
@@ -536,12 +553,351 @@ def inference_run(
+    show_default=True,
+    help="Maximum area size in square meters to include in the output. Disabled by default.",
+)
+@click.option(
+    "--close_interiors",
+    is_flag=True,
+    help="Remove the interiors holes in the polygons.",
+)
```
```suggestion
@click.option(
    "--close_interiors",
    is_flag=True,
    default=False,
    show_default=True,
    help="Remove the interiors holes in the polygons.",
)
```

m-mohr commented on ftw_tools/cli.py:
```diff
@@ -536,12 +553,351 @@ def inference_run(
+
+
+@inference.command(
+    "ftw-inference-instance-segmentation-all",
```
```suggestion
    "instance-segmentation-all",
```

isaaccorley commented on ftw_tools/download/download_img.py:
```diff
@@ -235,7 +235,7 @@ def query_stac(
 
 
-def create_input(win_a, win_b, out, overwrite, stac_host, bbox=None):
+def create_input(win_a, out, overwrite, stac_host, win_b=None, bbox=None):
```
This updates the download functionality to allow for only downloading a single image instead of requiring multiple images (winA AND winB)

isaaccorley commented on ftw_tools/models/baseline_inference.py:
```diff
@@ -121,12 +153,12 @@ def run(
         sampler=sampler,
         batch_size=batch_size,
-        num_workers=6,
+        num_workers=num_workers,
```
I removed the hardcoding of num_workers here to make it configurable from the CLI

calebrob6 commented:
needs `ultralytics` dependency somewhere

isaaccorley commented:
Weird. I can change how the code checks if it's empty instead I think by just using len(gdf)

calebrob6 commented:
yeah idk how changing the ultralytics version changed that

calebrob6 commented on ftw_tools/models/delineate_anything.py:
```diff
@@ -86,6 +87,10 @@ def pixel_to_geo(x, y, z=None):
+            return None
+
         df["geometry"] = df["segments"].apply(
             lambda x: shapely.geometry.Polygon(zip(x["y"], x["x"]))
```
should be x, y

calebrob6 commented:
a couple other random things:

- would be good to have the "run all the stuff" command take an output image and output file name (instead of just directory) and not save in parquet if the output file name ends in something else. Reason is I often want to save in GPKG and have all my output files in the same directory by name.
- Like we talked about on slack, I was playing around with upsampling on the fly and this seems to help performance -- if we add this as a command line option then we need to remember to scale the transform to make things work.
- I don't think `--image_size` does anything -- didn't really see an effect from changing it and it isn't explicitly set in the original repo.

isaaccorley commented on ftw_tools/models/utils.py:
```diff
@@ -0,0 +1,191 @@
+from fiboa_cli.parquet import create_parquet
+
+
+def merge_polygons(
```
This function seems a bit complicated but it's just finding overlapping pairs and then computing IoU and containment metrics and then merging any that exceed some threshold into a single polygon. This is because when patching with overlap we get duplicate predictions in the overlap regions that aren't filtered by NMS.

isaaccorley commented:
> a couple other random things:
> 
> * would be good to have the "run all the stuff" command take an output image and output file name (instead of just directory) and not save in parquet if the output file name ends in something else. Reason is I often want to save in GPKG and have all my output files in the same directory by name.
> * Like we talked about on slack, I was playing around with upsampling on the fly and this seems to help performance -- if we add this as a command line option then we need to remember to scale the transform to make things work.
> * I don't think `--image_size` does anything -- didn't really see an effect from changing it and it isn't explicitly set in the original repo.

- I'm following the other `ftw inference all` command which returns an output dir. I can update both but probably after this PR imo. I also updated to be able to save to parquet, gpkg, and geojson.
- I changed to use `resize_factor` instead of `image_size`. The inputs pixel coords are renormalized down to the patch size so no need to modify the transform.
- The image size issue is that if you pass a float tensor to ultralytics it ignores the imgsz parameter for whatever reason. This is now resized outside of the ultralytics model predict using torchvision.

isaaccorley commented:
> yeah idk how changing the ultralytics version changed that

I changed it to use `if len(df) == 0` instead of `if df.empty` so this is probably fixed now.

hannah-rae commented:
Sorry this took me so long to try @isaaccorley! 

Have you tested this on a Mac with `--mps_mode` enabled? I got this error:
```
NotImplementedError: The operator 'torchvision::nms' is not currently implemented for the MPS device. If you want this op to be considered for addition please comment on https://github.com/pytorch/pytorch/issues/141287 and mention use-case, that resulted in missing op as well as commit hash Unknown. As a temporary fix, you can set the environment variable `PYTORCH_ENABLE_MPS_FALLBACK=1` to use the CPU as a fallback for this op. WARNING: this will be slower than running natively on MPS.
```

To reproduce:
`ftw inference run-instance-segmentation austria.tif -o austria-delineateanything.gpkg --gpu 0 --mps_mode`

isaaccorley commented:
@hannah-rae I've added a fix and tested that it worked on my mac. Can you try again?

isaaccorley commented:
@codex can you fix the test errors

isaaccorley commented:
@hannah-rae I _think_ this should be good to try again now -- thanks for bearing with me

hannah-rae commented:
Now we have a new error 😬 

```
  File "/Users/hkerner/src/ftw/ftw-baselines/ftw_tools/models/baseline_inference.py", line 344, in run_instance_segmentation
    polygons.append(model.polygonize(pred, transform, crs))
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/hkerner/src/ftw/ftw-baselines/ftw_tools/models/delineate_anything.py", line 93, in polygonize
    df = result.to_df().to_pandas()
         ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/hkerner/src/ftw/ftw-baselines/.pixi/envs/dev/lib/python3.12/site-packages/pandas/core/generic.py", line 6318, in __getattr__
    return object.__getattribute__(self, name)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'DataFrame' object has no attribute 'to_pandas'
```
